backpack/core/derivatives/nll_base.py

```diff
@@ -0,0 +1,132 @@
+        self,
+        module: CrossEntropyLoss,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mc_samples: int = 1,
+        subsampling: List[int] = None,
+    ) -> Tensor:
+        """Method to approximate the square root Hessian through Monte Carlo sampling.
+
+        For a given loss, either _make_distribution or compute_sampled_grads must be
+        implemented.
+
+        Optionally, it is possible to specify _checks to perform before calculating
+        the loss gradient.
+
+        For use in mean mode, _get_mean_normalization must be implemented.
+
+        Args:
+            module:
+            g_inp:
+            g_out:
+            mc_samples: number of Monte Carlo samples to take
+            subsampling: Indices of samples that are sliced along the dimension.
+            Defaults to ``None`` (use all samples).
+
+        Returns:
+            Monte Carlo sampled gradients
+
+        """
+        self._checks(module)
+        subsampled_input = Variable(
+            subsample(module.input0, subsampling=subsampling), requires_grad=True
+        )
+        sampled_grads = self.compute_sampled_grads(subsampled_input, mc_samples) / sqrt(
+            mc_samples
+        )
+        if module.reduction == "mean":
+            sampled_grads /= sqrt(self._get_mean_normalization(module.input0))
+        return sampled_grads
+
+    def _checks(self, module):
+        """Any checks to be performed. Default runs none."""
+        return
+
+    def compute_sampled_grads(self, subsampled_input, mc_samples):
+        """Method to create the sampled gradients.
+
+        For loss functions that can be expressed as a Negative Log Likelihood (NLL), the
+        Maximum a-Posteriori (MAP) estimate of the network parameters can be written as
+        𝜃ₘₐₚ = argmin_𝜃 (𝑟(𝜃)+𝑙(xₙ,yₙ;𝜃)) where 𝑟 is the regularizer and 𝑙 is the loss
+        function, defined here as 𝑙(xₙ,yₙ;𝜃)= −log p(yₙ | f_𝜃(xₙ)). TODO how to cite this?
+
+        compute_sampled_grads calculates and returns the gradient of the loss with respect
+        to each of the randomly drawn samples. To use this function, the user must implement
+        the functions _make_distribution.
+
+        Alternatively, if a faster gradient calculation is known for the loss function, the
+        user may overwrite this method entirely with a hand-crafted method. Then the method
+        should return the gradient of the loss with respect to the subsampled input for each
+        of the Monte Carlo samples
+
+        Args:
+            subsampled_input: input after subsampling
+            mc_samples: number of samples
+
+        Returns:
+            sampled gradient of shape [mc_samples, *subsampled_input.shape]
+
+        """
+        with enable_grad():
+            gradient = []
+            dist = self._make_distribution(subsampled_input)
+            y_tilde = dist.sample(sample_shape=Size([mc_samples]))
+            loss_tilde = -dist.log_prob(y_tilde)
```

Comment:
I remember we discussed at some point whether this should be inside the loop over `mc_samples` or not.
I think we moved it outside in an attempt to make it work with `retain_graph=False`. Would it make sense to move it back in, given that we need to use `retain_graph=True` anyways?

This will change `y_tilde = dist.sample()` and `outputs=loss_tilde.sum()` and `for _ in range(mc_samples)`.

Reply:
Fixed in comment
@jabader97 [Fixed some commenting, removed mean from mse make distribution, moved…](https://github.com/f-dangel/backpack/pull/252/commits/97cb44f6a85d3bb5b7ec3de309d871fd18f67f8c)